client/verta/verta/_tracking/deployable_entity.py

```diff
@@ -0,0 +1,130 @@
+
+        # get paths to files within
+        if paths is None:
+            # Python files within filtered sys.path dirs
+            paths = local_sys_paths
+            extensions = ['py', 'pyc', 'pyo']
+        else:
+            # all user-specified files
+            paths = paths
+            extensions = None
+        local_filepaths = _utils.find_filepaths(
+            paths, extensions=extensions,
+            include_hidden=True,
+            include_venv=False,  # ignore virtual environments nested within
+        )
+
+        # obtain deepest common directory
+        #     This directory on the local system will be mirrored in `_CUSTOM_MODULES_DIR` in
+        #     deployment.
+        curr_dir = os.path.join(os.getcwd(), "")
+        paths_plus = list(local_filepaths) + [curr_dir]
+        common_prefix = os.path.commonprefix(paths_plus)
+        common_dir = os.path.dirname(common_prefix)
+
+        # replace `common_dir` with `_CUSTOM_MODULES_DIR` for deployment sys.path
+        depl_sys_paths = list(map(lambda path: os.path.relpath(path, common_dir), local_sys_paths))
+        depl_sys_paths = list(map(lambda path: os.path.join(_CUSTOM_MODULES_DIR, path), depl_sys_paths))
+
+        bytestream = six.BytesIO()
+        with zipfile.ZipFile(bytestream, 'w') as zipf:
+            for filepath in local_filepaths:
+                arcname = os.path.relpath(filepath, common_dir)  # filepath relative to archive root
+                try:
+                    zipf.write(filepath, arcname)
+                except:
+                    # maybe file has corrupt metadata; try reading then writing contents
+                    with open(filepath, 'rb') as f:
+                        zipf.writestr(arcname, f.read())
+
+            # add verta config file for sys.path and chdir
+            working_dir = os.path.join(_CUSTOM_MODULES_DIR, os.path.relpath(curr_dir, common_dir))
+            zipf.writestr(
+                "_verta_config.py",
+                six.ensure_binary('\n'.join([
+                    "import os, sys",
+                    "",
+                    "",
+                    "sys.path = sys.path[:1] + {} + sys.path[1:]".format(depl_sys_paths),
+                    "",
+                    "try:",
+                    "    os.makedirs(\"{}\")".format(working_dir),
+                    "except OSError:  # already exists",
+                    "    pass",
+                    "os.chdir(\"{}\")".format(working_dir),
+                ]))
+            )
+
+            # add __init__.py
+            init_filename = "__init__.py"
+            if init_filename not in zipf.namelist():
+                zipf.writestr(init_filename, b"")
+
+            if self._conf.debug:
+                print("[DEBUG] archive contains:")
+                zipf.printdir()
+        bytestream.seek(0)
+
+        self._log_modules_as_artifact(bytestream, overwrite=overwrite)
```

Comment:
Instead of adding another layer of function calls that can be harder to follow, do you think we could change this function to `_custom_modules_as_artifact()`, and have it return `bytestream` for the caller to log directly?

Let me know what you think. And if you're not totally comfortable with that just yet, we can make it a follow-up ticket to address next sprint.

Reply:
Great idea actually. I _think_ that I can get sth like that to happen. Check the updated code!